feat :리뷰 목록 조회 API 구현 (테스트 커버리지 100%)

ParkKangbin opened the pull request:
##  PR 요약
- 관련 이슈 번호: #110 
- 작업 요약:  스터디 그룹 리뷰 목록 조회 API를 구현했습니다.

## 📄 상세 내용
- ReviewListRequestSerializer: `group_uuid` 유효성 검증 및 queryset 반환
- ReviewListItemSerializer / ReviewListResponseSerializer 추가
  - rating, content, created_at 필드만 노출
- 최신순 정렬 기능 적용
- 리뷰가 없는 경우 404 응답 처리
- 테스트 코드 작성 및 테스트 커버리지 100% 달성
## 📸 스크린샷 (선택)
<img width="835" height="545" alt="image" src="https://github.com/user-attachments/assets/95483737-99cc-475f-8d03-fcc02258ac8d" />

## 📝 기타 참고 사항
- API 명세서와 응답 형식 일치
- 사용자 요구 정의서 반영
- 커밋메세지를 잘못 남겨 변경하려고 했으나 vim에서 style 이모지가 입력되지 않아 이모지를 넣지 못했습니다.
## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 모두 통과했습니다. (커버리지 100%)

ParkKangbin committed:
✨feat : 리뷰 목록 조회 API 구현

ParkKangbin committed:
style: 코드 포매팅 수정

SimJunSik commented on apps/studies/serializers/review_serializers.py:
```diff
@@ -64,3 +66,60 @@ class Meta:
+        self.context["study_group"] = group
+        return value
+
+    def get_queryset(self) -> QuerySet[StudyReview]:
+        # 검증된 StudyGroup에 속한 리뷰 목록을 최신순으로 조회, 없으면 404 반환
+        group: StudyGroup = self.context["study_group"]
+        qs = StudyReview.objects.filter(study_group=group).order_by("-created_at")
+        if not qs.exists():
+            raise NotFound("해당 스터디 그룹에 대한 리뷰가 존재하지 않습니다.")
+        return qs
+
+    def build_payload(self) -> Dict[str, Any]:
+        # 응답에 필요한 데이터(study_group_id, reviews) 구성
+        group: StudyGroup = self.context["study_group"]
+        reviews = self.get_queryset()
+        return {"study_group_id": group.id, "reviews": reviews}
```
Serializer 의 역할은 검증, 직렬화인데 NotFound 같은 로직이 여기 있는게 적절한지 다시 생각해보면 좋을 것 같습니다

ParkKangbin replied:
get_queryset() 안에서 NotFound를 던지는 대신 -> View 에서 queryset을 조회하고, 결과가 비어 있으면 빈 리스트 그대로 내려주도록 수정하겠습니다.

ParkKangbin replied:
없는 스터디 그룹은 리소스 자체가 존재하지 않는 상황이므로 404 NotFound가 맞다고 판단했고, 그래서 validate_group_uuid에서는 그대로 NotFound를 사용했습니다. 다만 필요하다면 ValidationError로 바꿔서 View에서 404 처리하도록 옮길 수도 있습니다

SimJunSik commented on apps/studies/serializers/review_serializers.py:
```diff
@@ -64,3 +66,60 @@ class Meta:
+        # 검증된 StudyGroup에 속한 리뷰 목록을 최신순으로 조회, 없으면 404 반환
+        group: StudyGroup = self.context["study_group"]
+        qs = StudyReview.objects.filter(study_group=group).order_by("-created_at")
+        if not qs.exists():
+            raise NotFound("해당 스터디 그룹에 대한 리뷰가 존재하지 않습니다.")
```
보통 filter 를 해서 결과가 없으면 empty list 를 반환하고, 에러를 반환하진 않습니다

ParkKangbin committed:
♻️ refactor : get_queryset()안에서 NotFOUND를 던지는 대신 View 에서queryset을 조회하고  결과가 비어있으면 빈 리스트 그대로 내려주도록 수정

Meoyoug commented on apps/studies/serializers/review_serializers.py:
django 에서 path parameter를 검증하는 것은 시리얼라이저에서 하지않습니다.
url 파싱할때 장고에서 처리해줍니다.

```
class ReviewListRequestSerializer(serializers.Serializer[Dict[str, Any]]):
    """
    스터디 그룹 리뷰 목록 조회 요청 전용 Serializer
    """
    group_uuid = serializers.UUIDField()
    def validate_group_uuid(self, value: Any) -> Any:
        # group_uuid가 유효한 StudyGroup인지 검증, 없으면 404 반환
        try:
            group = StudyGroup.objects.get(uuid=value)
        except StudyGroup.DoesNotExist as exc:
            raise NotFound(detail="해당 스터디 그룹이 존재하지 않습니다.") from exc
        self.context["study_group"] = group
        return value
```
이 부분에 대해서는 불필요한 2중 검증이되며, uuid를 통해 객체를 가져오는 부분은 service나 view내에서 구현하고 객체가 없을 시 404를 반환하는 것 역시 View에서 이루어져야합니다.

Meoyoug commented on apps/studies/views/review_views.py:
```diff
@@ -45,9 +46,14 @@ class StudyGroupReviewListView(APIView):
     permission_classes = [IsAuthenticated]
 
     def get(self, request: Request, group_uuid: UUID, *args: object, **kwargs: object) -> Response:
```
여기에서 차라리 이전에 지웠던 

```
    def get_queryset(self) -> QuerySet[StudyReview]:
        # 검증된 StudyGroup에 속한 리뷰 목록을 최신순으로 조회, 없으면 404 반환
        group: StudyGroup = self.context["study_group"]
        qs = StudyReview.objects.filter(study_group=group).order_by("-created_at")
        if not qs.exists():
            raise NotFound("해당 스터디 그룹에 대한 리뷰가 존재하지 않습니다.")
        return qs
```
이 코드를 View에서 메서드를 선언해서 사용하도록 해야합니다.

ParkKangbin replied:
path parameter 검증 로직을 Serializer에서 제거하고, View의 get_queryset에서 get_object_or_404로 처리하도록 변경했습니다.
리뷰가 없는 경우에는 404 대신 빈 배열을 반환하도록 수정했습니다.

ParkKangbin committed:
♻️ refactor: path parameter 검증 로직을 Serializer에서 제거하고 view의
get_queryset에서 get_object_or_404로 처리

Meoyoug commented on apps/studies/views/review_views.py:
```diff
@@ -45,14 +47,16 @@ class StudyGroupReviewListView(APIView):
-        group: StudyGroup = req.context["study_group"]
+    def get_queryset(self, group_uuid: UUID) -> Tuple[StudyGroup, "QuerySet[StudyReview]"]:
+        # 스터디 그룹 검증 (없으면 404)
+        group = get_object_or_404(StudyGroup, uuid=group_uuid)
```
여기부터도 쿼리 두번날릴필요없이

```
qs = StudyReview.objects.filter(study_group__uuid=group_uuid).order_by("-created_at")
```

ParkKangbin replied:
기존에는 get_object_or_404(StudyGroup, uuid=group_uuid)로 항상 스터디 그룹을 먼저 조회한 뒤, 해당 그룹의 리뷰를 다시 조회하는 방식이라 쿼리가 최소 2번 발생했습니다.)
피드백 주신 내용을 반영해 현재는 StudyReview.objects.filter(study_group__uuid=group_uuid)로 리뷰만 먼저 조회하도록 수정하여 리뷰가 존재할 경우 쿼리가 1번만 실행되도록 개선했습니다.
다만 API 명세상 study_group_id는 항상 응답에 포함되어야 해서 리뷰가 전혀 없는 경우에는 스터디 그룹 존재 여부 확인을 위해 StudyGroup을 한 번 더 조회하는 로직을 유지했습니다.
최대한 타협을 했다고 생각합니다 ㅠㅠ

Umdoong replied:
저희는 외부에 노출시킬 때는 id가 아니라 uuid를 노출시키기로 했습니다.
아마 명세서가 잘못 작성되어 있나보네요.
group_uuid의 응답이 필요하다면 프론트에서 넘겨준 데이터를 그대로 응답해주면 될 것 같습니다.

ParkKangbin committed:
♻️ refactor: 리뷰어 피드백 반영 - StudyGroupReviewListView 쿼리 최적화

- 기존: 항상 StudyGroup 조회 후 Review 조회 → 최소 2회 쿼리 발생
- 변경: Review에서 group_uuid 직접 조회 → 리뷰가 있는 경우 쿼리 1회로
  축소
  - 단, API 명세서 상 study_group_id 항상 응답 필요 → 리뷰 없을 시
    StudyGroup 조회 유지

Umdoong commented on apps/studies/urls/review_urls.py:
```diff
@@ -1,7 +1,11 @@
 
 urlpatterns = [
     path("/reviews", ReviewCreateAPIView.as_view(), name="review-create"),
+    path("/reviews/<uuid:group_uuid>/", StudyGroupReviewListView.as_view(), name="study-group-review-list"),
```
trailing slash 빼주세요!

Umdoong commented on apps/studies/views/review_views.py:
```diff
@@ -27,3 +36,31 @@ def post(self, request: Request, *args: object, **kwargs: object) -> Response:
+
+    permission_classes = [IsAuthenticated]
+
+    def get_queryset(self, group_uuid: UUID) -> QuerySet[StudyReview]:
```
재활용을 할 쿼리셋이면 서비스 레이어 or manager 에서 관리하시고
여기서만 쓸거라면 그냥 view 내에서 바로 orm을 사용시는 편이 더 좋아보여요

Umdoong commented on apps/studies/serializers/review_serializers.py:
```diff
@@ -64,3 +66,28 @@ class Meta:
+    스터디 그룹 리뷰 목록 응답 Envelope Serializer
+    """
+
+    study_group_id = serializers.IntegerField()
```
nasted serializer를 활용해보세요!

ParkKangbin committed:
♻️ refactor: 분리된 Serializer/View 구조로 리팩터링
